docs/proposals/20231227-koordlet-resctrl-qos-enhance.md

```diff
@@ -0,0 +1,462 @@
+
+### Goals
+
+- Migrate existing fixed class LLC and MBA configuration to NRI-powered runtime hooks for timely execution
+- Add LLC and MBA monitor for fixed class
+- Add pod level LLC and MBA configuration/adjustment and monitor
+
+### Non-Goals/Future Work
+
+- Resctrl policy to better use LLC and MBA resource
+- QoS manager plugin to detect noisy neighbor based on CPU, Memory, LLC, MBA and potential other resources to adjust LLC and MBA
+- Scheduler based on LLC and MBA resource
+
+## Proposal
+
+To achieve real-time control over LLC and MBA at fixed class and pod level, we will implement a runtime hook plugin dedicated to these functionalities. This plugin, named the Resctrl runtime hook plugin, will leverage NRI to facilitate timely adjustments and granular resource management.
+
+- Resctrl runtime hook will create/update QoS class level by using Rule to subscribe NodeSLO
+- Resctrl runtime hook subscribe RunPodSandBox event, and will handle pod level LLC and MBA schemata init
+- Resctrl runtime hook subscribe CreateContainer event, and set closid to ContainerContext, ContainerContext will update OCI spec based on the closid
+- Resctrl runtime hook will register a callback to Reconciler to consume ContainerTaskIds and then update new ContainerTaskIds to corresponding resctrl control group and monitor group
+
+Concurrently with the implementation of the Resctrl runtime hook plugin, we will deploy a dedicated Resctrl metric collector. This collector leverages NodeSLO and PodsInformer to gather real-time data on fixed class and pod-level LLC and MBA. This architecture, as depicted in the below diagram, ensures comprehensive resource consumption insights, which are crucial for informing the dynamic adjustments.
+![Alt text](../images/resctrl.png)
+
+
+### User Stories
+
+#### Story 1
+As a cluster administer, I want to apply and adjust LLC/MBA QoS class configuration during runtime to get better resource usage
+#### Story 2
+As a user, I want to adjust my workload's LLC/MBA resource during runtime.
+#### Story 3
+As a cluster administer, I want to monitor cluster LLC/MBA resource usage.
+#### Story 4
+As a cluster administer, I find some workloads are noisy neighbor, I want to limit these noisy neighbors' LLC/MBA.
+
+### Requirements (Optional)
+Need Koordinator to upgrade to 1.3.0+
```

Comment:
This proposal may be implemented in 1.4.x or 1.5.0 since 1.4.0 will be released soon.